src/modules/iotjs_module_i2c.c

```diff
@@ -157,132 +168,164 @@ void AfterI2CWork(uv_work_t* work_req, int status) {
+  }
 
   iotjs_jargs_destroy(&jargs);
   iotjs_i2c_reqwrap_dispatched(req_wrap);
 }
 
-static void GetI2cArray(const jerry_value_t jarray,
-                        iotjs_i2c_reqdata_t* req_data) {
-  // FIXME
-  // Need to implement a function to get array info from jerry_value_t Array.
-  jerry_value_t jlength =
-      iotjs_jval_get_property(jarray, IOTJS_MAGIC_STRING_LENGTH);
-  IOTJS_ASSERT(!jerry_value_is_undefined(jlength));
-
-  req_data->buf_len = iotjs_jval_as_number(jlength);
-  req_data->buf_data = iotjs_buffer_allocate(req_data->buf_len);
-
-  for (uint8_t i = 0; i < req_data->buf_len; i++) {
-    jerry_value_t jdata = iotjs_jval_get_property_by_index(jarray, i);
-    req_data->buf_data[i] = iotjs_jval_as_number(jdata);
-    jerry_release_value(jdata);
-  }
-
-  jerry_release_value(jlength);
-}
-
-#define I2C_ASYNC(op)                                                  \
+#define I2C_CALL_ASNCY(op, jcallback)                                  \
   do {                                                                 \
     uv_loop_t* loop = iotjs_environment_loop(iotjs_environment_get()); \
+    iotjs_i2c_reqwrap_t* req_wrap =                                    \
+        iotjs_i2c_reqwrap_create(jcallback, i2c, op);                  \
     uv_work_t* req = iotjs_i2c_reqwrap_req(req_wrap);                  \
-    uv_queue_work(loop, req, op##Worker, AfterI2CWork);                \
+    uv_queue_work(loop, req, i2c_worker, i2c_after_worker);            \
   } while (0)
 
 JS_FUNCTION(I2cCons) {
   DJS_CHECK_THIS();
+  DJS_CHECK_ARGS(1, object);
+  DJS_CHECK_ARG_IF_EXIST(1, function);
+
   // Create I2C object
   const jerry_value_t ji2c = JS_GET_THIS();
-#ifdef __linux__
-  DJS_CHECK_ARGS(2, string, function);
-  iotjs_string_t device = JS_GET_ARG(0, string);
-#else
-  DJS_CHECK_ARGS(2, number, function);
-  int device = JS_GET_ARG(0, number);
-#endif
-  iotjs_i2c_t* i2c = iotjs_i2c_create(&device, ji2c);
+  iotjs_i2c_t* i2c = iotjs_i2c_create(ji2c);
+  IOTJS_VALIDATED_STRUCT_METHOD(iotjs_i2c_t, i2c);
+
+  jerry_value_t jconfig;
+  JS_GET_REQUIRED_ARG_VALUE(0, jconfig, IOTJS_MAGIC_STRING_CONFIG, object);
 
-  IOTJS_ASSERT(i2c ==
-               (iotjs_i2c_t*)(iotjs_jval_get_object_native_handle(ji2c)));
+  jerry_value_t res = iotjs_i2c_set_platform_config(i2c, jconfig);
+  if (jerry_value_has_error_flag(res)) {
+    return res;
+  }
+
+  DJS_GET_REQUIRED_CONF_VALUE(jconfig, _this->address,
+                              IOTJS_MAGIC_STRING_ADDRESS, number);
 
-  // Create I2C request wrap
-  const jerry_value_t jcallback = JS_GET_ARG(1, function);
-  iotjs_i2c_reqwrap_t* req_wrap =
-      iotjs_i2c_reqwrap_create(jcallback, i2c, kI2cOpOpen);
+  jerry_value_t jcallback = JS_GET_ARG_IF_EXIST(1, function);
 
-  I2C_ASYNC(Open);
+  // If the callback doesn't exist, it is completed synchronously.
+  // Otherwise, it will be executed asynchronously.
+  if (jerry_value_is_null(jcallback)) {
```

Comment:
Maybe it would better to write it like this:
 
```c
if (!jerry_value_is_null(jcallback)) {
  I2C_CALL_ASNCY(kI2cOpOpen, jcallback);
} else if (!iotjs_i2c_open(i2c)) {
  return JS_CREATE_ERROR(COMMON, "I2C Error: cannot open I2C");
}
```